File: job-offer/src/main/java/com/oxygensend/joboffer/infrastructure/kafka/KafkaEventPublisher.java
package com.oxygensend.joboffer.infrastructure.kafka;

import com.oxygensend.joboffer.application.properties.JobOffersProperties;
import com.oxygensend.joboffer.domain.event.DomainEvent;
import com.oxygensend.joboffer.domain.event.DomainEventPublisher;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

@Component
final class KafkaEventPublisher implements DomainEventPublisher {
    private final KafkaTemplate<String, DomainEvent> kafkaTemplate;
    private final JobOffersProperties jobOffersProperties;

    KafkaEventPublisher(KafkaTemplate<String, DomainEvent> kafkaTemplate, JobOffersProperties jobOffersProperties) {
        this.kafkaTemplate = kafkaTemplate;
        this.jobOffersProperties = jobOffersProperties;
    }

    @Override
    public void publish(DomainEvent domainEvent) {
        kafkaTemplate.send(jobOffersProperties.topics().get(domainEvent.topic()), domainEvent.key(), domainEvent);
    }
}

File: opinions/src/main/kotlin/com/oxygensend/opinions/infrastructure/kafka/KafkaEventPublisher.kt
package com.oxygensend.opinions.infrastructure.kafka

import com.oxygensend.opinions.domain.event.DomainEvent
import com.oxygensend.opinions.domain.event.DomainEventPublisher
import com.oxygensend.opinions.domain.event.Topics
import org.springframework.kafka.core.KafkaTemplate

internal class KafkaEventPublisher(
    private val kafkaTemplate: KafkaTemplate<String, DomainEvent>,
    private val topics: Map<Topics, String>
) : DomainEventPublisher {

    override fun publish(domainEvent: DomainEvent) {
        topics[domainEvent.topic()]?.let { kafkaTemplate.send(it, domainEvent.key(), domainEvent) };
    }
}